### ISSUE TITLE
Improve GPU testing

### ISSUE THREAD

**Issue by makortel** [cms-sw/cmssw#29313]
Tests of PR https://github.com/cms-sw/cmssw/pull/29110 revealed that in the current testing infrastructure
* GPU drivers are too old, so code using CUDA 10.2 can not run
* All our GPU tests report success if they can't run on GPU to not report false positives on CPU-only machines. The tests also report success on GPU-equipped machines if the CUDA runtime can not be initialized, and that hid the former issue.

**Comment by makortel** [comment]
assign core

**Comment by makortel** [comment]
> * All our GPU tests report success if they can't run on GPU to not report false positives on CPU-only machines. The tests also report success on GPU-equipped machines if the CUDA runtime can not be initialized, and that hid the former issue.

I'm thinking if we should add a mechanism to force the GPU usage instead of the "auto discovery" in specific cases. An environment variable would be a straightforward way to propagate the "forcing" to all unit tests and `SwitchProducerCUDA` (even though in general I'd prefer more explicit method).

**Comment by makortel** [comment]
assign heterogeneous

**Comment by smuzaffar** [comment]
agree, scram can set an enviroment variable if GPU is available and should be used. We do not need all unit tests and SwitchProducerCUDA to check for this, just having an extra unit test which check for this env variable and make sure that it can find and use the deice is good enough.

**Comment by makortel** [comment]
Good point, one test should be sufficient to test that the GPU tests are really run on GPUs or not (and then the environment variable use wouldn't spread)

**Comment by fwyzard** [comment]
I'm not sure it's a good idea to allow forcing a `SwitchProducer` (either via an environment variable or other methods), because it opens a lot of doors for abuse downstream.

For example,  oneAPI is currently using this approach to force a backend (for the default device selector) and IMHO is running into tons of issues with it.

I have no objections to an override for the unit tests, though.

**Comment by fwyzard** [comment]
> agree, scram can set an environment variable if GPU is available and should be used.

How does SCRAM know that ?

**Comment by smuzaffar** [comment]
@fwyzard , the same was scram set the cuda env ( https://github.com/cms-sw/cmssw-config/blob/master/SCRAM/hooks/runtime/00-nvidia-drivers#L20 ) . If scram finds that system have nvidia then it sets the env variable .

**Comment by fwyzard** [comment]
OK, what you are saying is that if there is an NVIDIA GPU with a loaded driver, you assume that we want to run one specific test using the GPU ?

**Comment by smuzaffar** [comment]
yes

**Comment by fwyzard** [comment]
OK... then it might be simpler to write a C++ or Python test that checks if there is a GPU and tries to use it, without involving SCRAM.

**Comment by smuzaffar** [comment]
sounds good.

**Comment by makortel** [comment]
@smuzaffar Replying here to https://github.com/cms-sw/cmssw/pull/29110#issuecomment-604479563

> worrying part is that these resources are not under our control so I can not guarantee that GPU tests will end of a GPU system which our software can use.

Is there any way to e.g. request a specific CUDA runtime version in the job submission?

If not, how about we run the test @fwyzard suggested in https://github.com/cms-sw/cmssw/issues/29313#issuecomment-604461920 always when we run any GPU tests? (until a better solution is found out)

**Comment by smuzaffar** [comment]
May be I can use "CUDADriverVersion >= 10.2" when I request the GPU resource. I see that only one GPU resources has CUDADriverVersion=10.1 and rest have 10.2

**Comment by fwyzard** [comment]
Yes, that looks like a good parameter to request.

**Comment by makortel** [comment]
> I'm not sure it's a good idea to allow forcing a `SwitchProducer` (either via an environment variable or other methods), because it opens a lot of doors for abuse downstream.

Could you elaborate your concerns? I would think in the long run we do want to control the devices to use externally to cmsRun (e.g. we run a full-node pilot and on a multi-GPU node want to run a process per GPU, or in a single-GPU node want to run CPU-only variant as well).

**Comment by fwyzard** [comment]
My feeling is that if we want to force a module to run on a CPU or a GPU, we should not put a SwitchProducer in the job.

If the idea is to limit the job to a subset of the resources, I think both CUDA and SYCL allow to do it via environment variables.
At least for CUDA this is also be done (and more reliably) via cgroups, limiting which devices the job has access to (similar to how taskset can be used to limit the CPU cores).

If the idea is to force individual modules to run on a CPU or GPU, an environment variable will not work.

**Comment by fwyzard** [comment]
If we want to be able to use a SwitchProducer to have a single configuration, but allow overriding the choice globally or on a module-by-module basis, naively I would find it more natural to set the override in python.

**Comment by fwyzard** [comment]
But, I admit I have not given too much thoughts to the details.

**Comment by makortel** [comment]
I meant controlling the behavior of the whole cmsRun (sorry if that was not clear). Yes, CUDA etc. provide means to control the resources, and it could be that those are sufficient, I don't know. Personally I would also prefer a python-based general solution for specifying the "allowed resources to use". 

I fully agree that forcing individual modules to run here or there is not for SwitchProducer, and would not even work in general unless the configuration contains modules to transfer and possibly convert data formats between here and there. I also think such a capability would not have much real value beyond testing.

**Comment by fwyzard** [comment]
+heterogeneous

I think the current alpaka-based approach and testing infrastructure address this issue ?

**Comment by makortel** [comment]
> I think the current alpaka-based approach and testing infrastructure address this issue ?

Probably

**Comment by makortel** [comment]
+core
